### DISCUSSION TITLE
这里是讨论区

### DISCUSSION THREAD

**Discussion by ruanqizhen** [ruanqizhen/labview_book#7] [General]
有关书中的具体问题，可以在每个章节下的留言区讨论。其它的问题可以新开一帖讨论。

**Comment by ZhaoLi-an** [comment]
https://ruanqizhen.wordpress.com/ ，这个博客文章无法评论，提交就会卡掉，是搭建问题还是网络问题？

**Reply by ruanqizhen** [reply]
应该不是搭建的问题，不是我自己搭建的，我用的是wordpress的免费服务。最有可能是网络问题。
如果那里连接不好，有啥问题也可以在这里讨论。

**Comment by Alvin2110** [comment]
源代码下载点开链接，显示”无法访问此网站“，请问是不是翻墙才能下载？

**Reply by ruanqizhen** [reply]
是这个连接吗： https://github.com/ruanqizhen/labview_book/tree/main/code
是只有这个无法打开，还是所有 github 上的项目都打不开？

**Reply by Alvin2110** [reply]
只有download那一步打不开，展开目录都没问题。

**Reply by ruanqizhen** [reply]
这个看起来不像是网络的问题，更像是你的某些系统设置引起的。你用一个其它的浏览器或者电脑试试看呢。或者尝试用GitHub提供的工具、命令行等试试看能不能下载。

**Comment by Alvin2110** [comment]
谢谢阮老师！我试试看。
祝好！



---- 回复的原邮件 ----
| 发件人 | ***@***.***> |
| 日期 | 2023年03月19日 22:24 |
| 收件人 | ***@***.***> |
| 抄送至 | ***@***.***>***@***.***> |
| 主题 | Re: [ruanqizhen/labview_book] 这里是讨论区 (Discussion #7) |

这个看起来不像是网络的问题，更像是你的某些系统设置引起的。你用一个其它的浏览器或者电脑试试看呢。或者尝试用GitHub提供的工具、命令行等试试看能不能下载。

**Comment by mastercs1** [comment]
先生您好，
十分冒昧 想请教您一个问题，最近在拜读您的《labview 编程经验》，因为本人是一个文本编程人员，所以有些问题还是有些不明白，
在这里向您提问。

在移位寄存器中，我理解 您的意思是一对移位寄存器是一个 内存地址（可能是我理解错误）。 那么，当把一个数放入到，输入端的时候，输出端也就同时获得了这个值。那么在您给的例子当中，

![download](https://github.com/ruanqizhen/labview_book/assets/20642091/330cac7d-36d3-4482-83bb-cfb83409ca95)

为什么 在循环中，这一对寄存器还要连线呢？ 如果是同一内存地址，不连线 输出端不是也一定有值吗？

但是不连线， 程序会报错，那是为什么呢？

刚开始学习labview 我的问题 应该十分幼稚，如果能得到您的回答，我将不胜荣幸。

十分感谢您的教程。

**Reply by mastercs1** [reply]
我想我现在有一点明白了， 两个寄存器并不是同一地址，是相互独立的，但是连线后，就是两个地址的值迭代拷贝。两个寄存器在循环还没有判断是否运行前，系统就初始化了两个寄存器，因为连线 输入寄存器的值 拷贝到 输出寄存器。 这样理解对吗？总还是感觉怪怪的。
还是怀念 temp value. ...哈哈

**Reply by ruanqizhen** [reply]
可以这么理解：这一对寄存器对应的就是一个内存地址。数据从左边输入，右边也应该看到同样的数据。但是，它右边的这个是有一个数据输入接线端的，每次运行到它，必须输入一个数据。所以，如果不希望内存中的数据变化，就只能把它读出来再写回去。

有些地方为了防止出错，LabVIEW 限制的比较死。比如移位寄存器这个东西，数据通常是会在循环内改变的，如果不改变，也没有必要使用它了。所以，它就强制要求为右边寄存器的接线端连接数据线，防止用户忘记把改变后的数据存回去。

**Reply by mastercs1** [reply]
谢谢先生，现在理解了， 您说的对。以后不会的还要多请教您，祝您安好

**Comment by mastercs1** [comment]
先生请先原谅我吹毛求疵了。
![image](https://github.com/ruanqizhen/labview_book/assets/20642091/2e76d61f-0541-4a3e-ac2f-3e9c16f31afd)
图上的时间格式应该针对windows系统。在macos 上的 时间戳 与windows 格式不相同。
在 mac os 中 应该是
![image](https://github.com/ruanqizhen/labview_book/assets/20642091/12e229fb-a932-4545-af66-26a849631440)
我觉得如果兼顾多个系统，也许可以考虑在前面加一块 统一时间格式的代码。
![image](https://github.com/ruanqizhen/labview_book/assets/20642091/03957aab-54f9-40e4-b2f6-e73d8e2f1985)

班门弄斧，如果说的不对请您 指导批评。
另外，你这本大作 是有没有 paper based ?  如果有，是否可告知 如何购买？ 

十分感谢

**Reply by ruanqizhen** [reply]
你的建议非常好！我自己几乎没有使用过 MacOS，之前还真不了解它们的不同。我会根据你的建议，对这部分内容进行修改。

这本书很早以前曾经印刷出版过，但已经售罄，现在无法再买到纸质版的书籍了。

**Comment by samthephantom** [comment]
阮先生您好，请教一个问题，我们用labview 2018进行开发，想将程序发布成windows服务，以便程序自启以及被其他程序管理，我们之前是用了srvany.exe，在windows 10平台上运行，但是程序一天可能会崩溃一到两次，或者是电脑启动->程序启动的时候，可能初始化到一定程度（已经开始采集数据了）突然服务被关闭并且无法触发设备看门狗。
我们做了以下两种检查：
1、去看windows的系统日志，会有我们程序的APPCRASH，来自lvrt.dll。
2、查看用户\Documents\LabVIEW Data\LVInternalReports\， 寻求看labview内部报错的时候，文件夹里是空的。

请问您有遇到过这种状况吗？我们同事怀疑是srvany原因，也有可能是因为CPU占用太多（我们程序会对7到8个很长的数据做快速傅里叶变换所以CPU占用接近100%，每次持续大概10秒左右，每20秒一次）

**Reply by ruanqizhen** [reply]
你好，我没有做过类似的操作，也不清楚会是什么引起的问题。不过可以提供一些思路：可以先确定一下是不是和 srvany.exe 有关，试一下不通过 srvany.exe，直接启动 LabVIEW 程序，处理 FFT，看看是不是也有同样的问题。另外看一下内存使用情况，有没有内存泄露之类的问题。

**Reply by samthephantom** [reply]
您好，刚刚测试过不通过srvany.exe，直接通过打包生成的可执行文件启动Labview程序，有三次程序退出的情况 ，但只有一次提示“完成该操作所需内存不足”，其余没有error弹窗。这种情况在同样硬件参数的其他设备上比较少出现，因为我们有不少同规格的设备在运行同样的程序，都能运行很长时间，几个月内没有崩溃。

**Reply by ruanqizhen** [reply]
如果程序在其它硬件设备上运行都没有问题，那可能最简单的解决办法就是换硬件了。某个硬件质量不好，不稳定也很难说的。
